src/storage/recovery/recovery_manager.cpp

```diff
@@ -0,0 +1,762 @@
+    if (next_record->RecordType() == LogRecordType::REDO) {  // next record is a redo record
+      auto *next_redo_record = next_record->GetUnderlyingRecordBodyAs<RedoRecord>();
+      if (next_redo_record->GetDatabaseOid() == delete_record->GetDatabaseOid() &&
+          next_redo_record->GetTableOid() == delete_record->GetTableOid() &&
+          IsInsertRecord(next_redo_record)) {  // next record is an insert into the same pg_class
+        // Step 3: Get the oid and name for the database being created
+        auto pr_map = pg_database->ProjectionMapForOids(GetOidsForRedoRecord(pg_database, next_redo_record));
+        TERRIER_ASSERT(pr_map.find(catalog::DATOID_COL_OID) != pr_map.end(), "PR Map must contain class oid");
+        TERRIER_ASSERT(pr_map.find(catalog::DATNAME_COL_OID) != pr_map.end(), "PR Map must contain class name");
+        auto next_db_oid = *(reinterpret_cast<catalog::db_oid_t *>(
+            next_redo_record->Delta()->AccessWithNullCheck(pr_map[catalog::DATOID_COL_OID])));
+
+        // If the oid matches on the next record, this is a renaming
+        if (db_oid == next_db_oid) {
+          // Step 4: Extract out the new name
+          VarlenEntry name_varlen = *(reinterpret_cast<VarlenEntry *>(
+              next_redo_record->Delta()->AccessWithNullCheck(pr_map[catalog::DATNAME_COL_OID])));
+          std::string name_string(name_varlen.StringView());
+
+          // Step 5: Rename the database
+          auto result UNUSED_ATTRIBUTE = catalog_->RenameDatabase(txn, next_db_oid, name_string);
+          TERRIER_ASSERT(result, "Renaming of database should always succeed during replaying");
+
+          // Step 6: Update metadata and clean up additional record processed. We need to use the indexes on
+          // pg_database to find what tuple slot we just inserted into. We get the new tuple slot using the oid
+          // index.
+          auto pg_database_oid_index = catalog_->databases_oid_index_;
+          auto pr_init = pg_database_oid_index->GetProjectedRowInitializer();
+          buffer = common::AllocationUtil::AllocateAligned(pr_init.ProjectedRowSize());
+          pr = pr_init.InitializeRow(buffer);
+          *(reinterpret_cast<uint32_t *>(pr->AccessForceNotNull(0))) = static_cast<uint32_t>(next_db_oid);
+          std::vector<TupleSlot> tuple_slot_result;
+          pg_database_oid_index->ScanKey(*txn, *pr, &tuple_slot_result);
+          TERRIER_ASSERT(tuple_slot_result.size() == 1, "Index scan should only yield one result");
+          tuple_slot_map_[next_redo_record->GetTupleSlot()] = tuple_slot_result[0];
+          delete[] buffer;
+          tuple_slot_map_.erase(delete_record->GetTupleSlot());
+          delete[] reinterpret_cast<byte *>(next_redo_record);
+
+          return 1;  // We processed an additional record
+        }
+      }
+    }
+  }
+
+  // Step 3: If it wasn't a renaming, we simply need to drop the database
+  auto result UNUSED_ATTRIBUTE = catalog_->DeleteDatabase(txn, db_oid);
```

Comment:
How so? I thought DDL changes should not be visible to other concurrent transactions, or should cause one txn to abort. Unless you're talking about something else

Reply:
The problem is that your current method has no running transactions as you are consuming log records looking for a commit.  Since there are no running transactions, the GC is free to garbage collect _everything_ to include physical objects like indexes and tables even though there may be transactions for which you've seen "begin" records still waiting for a "commit" record.

Example:
```
T1               T2
                 BEGIN
BEGIN
DROP table1
COMMIT

[GC runs several times]

                 INSERT a INTO table1
                 INSERT b INTO table2
                 COMMIT
```

In the original database, T2 prevented the GC from deallocating the table or unlinking the catalog records until it exits.  Your replay, however, would currently let the GC happily destroy the `table1` forcing `T2` to now abort causing recovery to either fail or lose data (`b` in `table2`).

Reply:
T1 dropped the table before T2 inserted into it though, with a transactional catalog isn't that a write-write conflict so T2 should abort?

Reply:
No, T2 only reads the catalog so there is no conflict and it will proceed.

Reply:
Spoke with Andy. Firstly because we don't log BEGIN records, and we don't log until we fill a 4K byte redo buffer, the current implementation is probably not correct. Consider this schedule:
```
T1               T2
                 BEGIN
BEGIN
UPDATE A
COMMIT
                 UPDATE A
                 COMMIT
```

On the log, this will look like:
```
BEGIN T1
UPDATE A
COMMIT T1
BEGIN T2
UPDATE A
COMMIT T2
```
The current recovery manager will fully execute T1 before T2, which means T2 will get a snapshot AFTER T1 executed, and not the snapshot it should see BEFORE T1 executed.

The solution we discussed is logging all other active txns on the commit record of a txn. This will allow us at recovery time to wait to execute a txn until other txns with lower timestamps have executed. I will also need some sort of way to always "persist" which txns aborted, because this "active txn set" can possibly contain txns that aborted, but never flushed an abort record.